### ISSUE TITLE
Unable to do transfer tensors to GPU using `.cuda()` when using multiprocessing.Process with `fork`

### ISSUE THREAD

**Issue by Amir-Arsalan** [pytorch/pytorch#13883]
I know there has were issues regarding this last year but I thought it should have got resolved by now after about a year. I wonder, has there been any updates on this? I want to be able to use `fork` instead of `spawn` for running my processes as my work depends heavily on the benefits that `fork`ing provides.  

I just noticed that I cannot transfer my tensors to GPU when in a `fork`ed process from the `multiprocessing.Process` module. I do not even do `import torch` before I call `multiprocessing.Process` process via the `.start()` function but I get the following error:

 ```
File "/usr/local/lib/python3.6/dist-packages/torch/cuda/__init__.py", line 162, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (3) : initialization error at /pytorch/aten/src/THC/THCGeneral.cpp:51
```

So basically: 1) I run the process 2) do `import torch` inside of the `fork`ed process 3) call `torch.manual_seed()` and `torch.cuda.manual_seed()` functions inside the process 4) convert a Numpy tensor to a PyTorch tensor 5) transfer the PyTorch tensor to GPU via the `.cuda()` function. 6) The above error is thrown

Update: 
I noticed that some of the other packages that I have to import earlier on are probably calling the CUDA drivers and that's potentially why I am facing this issue. The strange thing is that although I manually called `torch.cuda.manual_seed()` before `fork`ing a process things are still find and I can do `.cuda()` but I cannot do this once I import the other packages in my workspace. I cannot change the pipeline of my work in terms of importing those packages and using them before doing `import torch` and I really need to find a way around this. I really hope there is gonna be a workaround for this. @ngimel @soumith

**Comment by soumith** [comment]
@Amir-Arsalan NVIDIA's CUDA stack is still not fork-safe. I dont think they will make it fork-safe anytime in the future either. There's not much we can actually do here as a PyTorch developer.

**Comment by ezyang** [comment]
What's the reason you can't avoid importing those packages before you fork? Would be interesting to know your constraint here.

**Comment by Amir-Arsalan** [comment]
@ezyang I have to generate data on the fly and then feed it to my models. I also have to do some pre-processing on some other stuff (3D objects in this case) before even generating the data.

**Comment by soumith** [comment]
repeating a message from Olexa on slack, closing this as wontfix::

```
@Amir, when I wrote in #general that CUDA is totally incompatible with forking, I meant this in the most serious and literal sense of those words possible.

This simply *cannot* be fixed in CUDA, PyTorch or anyone else. The problems inherent in `fork()`'ing *any* _multithreaded_ program are fundamentally unsolvable, and simply beyond the power of anyone to fix, at least not until a revolution in OS design happens.

I can only plead that you accept the general difficulty of safely forking within a multithreaded program and try to do things another way.

There's several blogs out there discussing why it's dangerous to fork in a multithreaded program, such as https://thorstenball.com/blog/2014/10/13/why-threads-cant-fork/

The gist of it is that when a thread `fork()`'s, *in the child process, all other threads _die instantly._* It doesn't matter what they were doing, they're _gone_.

- If they had locked a mutex, that mutex will never be unlocked again.
- If they were modifying a data structure, that data structure might be invalid.
- If they `malloc()`'ed some memory, that memory might never be deallocated, and `malloc()` may use locks and data structures anyways.
- If a thread was doing useful work, that work will never be complete, because the threads no longer exist.
- You can't join these non-existent threads.
- `pthread_atfork()` is a function meant to solve the problems above, but it's simply incapable of doing what it was meant to do safely, and that's why the POSIX.1 standard explicitly says that this function may be formally deprecated in the next version of the standard. It was a mistake.

So almost the only safe thing to do if you `fork()`'ed from a multi-threaded process is to call `exec()`. That's what `spawn` does.

Because the CUDA runtime uses threads to implement its runtime and asynchronous streams, once the CUDA runtime is initialized, it's insanely dangerous to `fork()`. Don't do it.
```

**Comment by Amir-Arsalan** [comment]
I sent an email to Ian Buck of NVIDIA about this issue. He just got back to me and said he has opened an internal tracker for the forking issue.

I hope they can somehow resolve this issue soon.
